drivers/storage/ide/pciidex/pciidex.h

```diff
@@ -1,74 +1,162 @@
 #ifndef _PCIIDEX_PCH_
 #define _PCIIDEX_PCH_
 
-#include <ntifs.h>
+#include <ntddk.h>
+#include <ntstrsafe.h>
+#include <initguid.h>
+#include <wdmguid.h>
 #include <ide.h>
 
-typedef struct _PCIIDEX_DRIVER_EXTENSION
-{
-	PCONTROLLER_PROPERTIES HwGetControllerProperties;
-	ULONG MiniControllerExtensionSize;
-	PCIIDE_UDMA_MODES_SUPPORTED HwUdmaModesSupported;
-} PCIIDEX_DRIVER_EXTENSION, *PPCIIDEX_DRIVER_EXTENSION;
+#include "bmdma.h"
+
+#define TAG_PCIIDEX    'XedI'
+
+#define IS_FDO(p)    (((PCOMMON_DEVICE_EXTENSION)(p))->IsFDO)
+
+#define IS_PRIMARY_CHANNEL(PdoExtension)    (PdoExtension->Channel == 0)
+
+typedef struct _PDO_DEVICE_EXTENSION    PDO_DEVICE_EXTENSION, *PPDO_DEVICE_EXTENSION;
 
 typedef struct _COMMON_DEVICE_EXTENSION
 {
-	BOOLEAN IsFDO;
+    BOOLEAN IsFDO;
+    PDEVICE_OBJECT Self;
+
+    _Write_guarded_by_(_Global_interlock_)
+    volatile LONG PageFiles;
+
+    _Write_guarded_by_(_Global_interlock_)
+    volatile LONG HibernateFiles;
+
+    _Write_guarded_by_(_Global_interlock_)
+    volatile LONG DumpFiles;
 } COMMON_DEVICE_EXTENSION, *PCOMMON_DEVICE_EXTENSION;
 
 typedef struct _FDO_DEVICE_EXTENSION
 {
-	COMMON_DEVICE_EXTENSION Common;
-
-	PBUS_INTERFACE_STANDARD BusInterface;
-	IDE_CONTROLLER_PROPERTIES Properties;
-	PHYSICAL_ADDRESS BusMasterPortBase;
-	PDEVICE_OBJECT LowerDevice;
-	PDEVICE_OBJECT Pdo[MAX_IDE_CHANNEL];
-	USHORT VendorId;
-	USHORT DeviceId;
-	PUCHAR MiniControllerExtension[0];
+    COMMON_DEVICE_EXTENSION Common;
+    PDEVICE_OBJECT Ldo;
+
+    ULONG Flags;
+
+    ULONG ControllerNumber;
+    BOOLEAN InNativeMode;
+    BOOLEAN IoBaseMapped;
+    BOOLEAN MiniportStarted;
+
+    PUCHAR BusMasterPortBase;
+
+    BUS_INTERFACE_STANDARD BusInterface;
+
+    KEVENT DeviceSyncEvent;
+    _Guarded_by_(DeviceSyncEvent)
+    PPDO_DEVICE_EXTENSION Channels[MAX_IDE_CHANNEL];
+
+    USHORT VendorId;
+    USHORT DeviceId;
+
+    PKINTERRUPT InterruptObject;
+    ULONG InterruptVector;
+    KIRQL InterruptLevel;
+    KINTERRUPT_MODE InterruptMode;
+    BOOLEAN InterruptShared;
+    KAFFINITY InterruptAffinity;
+
+    PDRIVER_OBJECT DriverObject;
+
+    KSPIN_LOCK BusDataLock;
+    IDE_CONTROLLER_PROPERTIES Properties;
+    PCONTROLLER_OBJECT ControllerObject;
+
+#ifdef _WIN64
+    DECLSPEC_ALIGN(8)
+#else
+    DECLSPEC_ALIGN(4)
+#endif
```

Comment:
This is not needed, you're putting a natural alignment for a pointer here

Reply:
I thought this will be the same with pciidex
https://github.com/reactos/reactos/commit/66cf1d20f4894017a5cc2ce9aaa1190b61854718